File: tobas_fake_publishers/src/tobas_fake_publishers/fake_gnss_publisher.py
import rospy

from tobas_tools_py.constants import Topic
from tobas_msgs.msg import Gps


class FakeGNSSPublisher:
    DEFAULT_SAMPLING_PERIOD = 0.2  # [s]
    DEFAULT_POS_STDDEV = 3.0  # [m]
    DEFAULT_VEL_STDDEV = 0.3  # [m/s]

    def __init__(self) -> None:
        sampling_period = rospy.get_param("~sampling_period", self.DEFAULT_SAMPLING_PERIOD)
        pos_stddev = rospy.get_param("~position_stddev", self.DEFAULT_POS_STDDEV)
        vel_stddev = rospy.get_param("~velocity_stddev", self.DEFAULT_VEL_STDDEV)

        pos_var = pos_stddev**2
        vel_var = vel_stddev**2

        self._gnss_msg = Gps()
        self._gnss_msg.fix_type = Gps.FIX_3D
        self._gnss_msg.latitude = 0.0
        self._gnss_msg.longitude = 0.0
        self._gnss_msg.altitude = 0.0
        self._gnss_msg.ground_speed.x = 0.0
        self._gnss_msg.ground_speed.y = 0.0
        self._gnss_msg.ground_speed.z = 0.0
        self._gnss_msg.position_covariance = [pos_var, 0, 0, 0, pos_var, 0, 0, 0, pos_var]
        self._gnss_msg.velocity_covariance = [vel_var, 0, 0, 0, vel_var, 0, 0, 0, vel_var]

        self._gnss_pub = rospy.Publisher(Topic.GNSS, Gps, queue_size=1)
        self._timer = rospy.Timer(rospy.Duration(sampling_period), self._timer_cb)

    def _timer_cb(self, event: rospy.timer.TimerEvent) -> None:
        self._gnss_msg.header.stamp = event.current_real
        self._gnss_pub.publish(self._gnss_msg)

File: tobas_fake_publishers/nodes/fake_gnss_publisher.cpp
#include <tobas_node/node.hpp>
#include <tobas_constants/constants.hpp>
#include <tobas_msgs_adapter/gps.hpp>

using namespace std;

class FakeGNSSPublisherNode : public tobas::BaseNode
{
  static constexpr auto kSamplingPeriod = 200ms;

  static constexpr double kDefaultPosStddev = 3.;   // [m]
  static constexpr double kDefaultVelStddev = 0.3;  // [m/s]

  using self = FakeGNSSPublisherNode;
  using super = tobas::BaseNode;

public:
  explicit FakeGNSSPublisherNode(const rclcpp::NodeOptions& options = rclcpp::NodeOptions());

private:
  double pos_stddev_;
  double vel_stddev_;

  ros2::PublisherPtr<tobas_msgs::Gps> gnss_pub_;
  ros2::TimerPtr timer_;

  void timerCb();
};

FakeGNSSPublisherNode::FakeGNSSPublisherNode(const rclcpp::NodeOptions& options)
  : super("fake_battery_publisher", options)
{
  pos_stddev_ = getDoubleParam("position_stddev", kDefaultPosStddev);
  vel_stddev_ = getDoubleParam("velocity_stddev", kDefaultVelStddev);

  gnss_pub_ = createPublisher<tobas_msgs::Gps>(tobas::kGNSSTopic);
  timer_ = createTimer(kSamplingPeriod, &self::timerCb, this);
}

void FakeGNSSPublisherNode::timerCb()
{
  auto gnss_msg = std::make_unique<tobas_msgs::Gps>();
  gnss_msg->header.stamp = get_clock()->now();
  gnss_msg->fix_type = tobas_msgs::msg::Gps::FIX_3D;
  gnss_msg->latitude = 0.;
  gnss_msg->longitude = 0.;
  gnss_msg->altitude = 0.;
  gnss_msg->ground_speed.setZero();
  gnss_msg->position_covariance = Eigen::Vector3d::Constant(pos_stddev_).asDiagonal();
  gnss_msg->velocity_covariance = Eigen::Vector3d::Constant(vel_stddev_).asDiagonal();

  gnss_pub_->publish(move(gnss_msg));
}

RCLCPP_COMPONENTS_REGISTER_NODE(FakeGNSSPublisherNode)
